File: src/test/java/br/com/jiratorio/domain/impediment/calculator/ImpedimentCalculatorByFlagTest.java
package br.com.jiratorio.domain.impediment.calculator;

import br.com.jiratorio.domain.changelog.JiraChangelogItem;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class ImpedimentCalculatorByFlagTest {

    private ImpedimentCalculatorByFlag calculator;

    @BeforeEach
    void setUp() {
        calculator = new ImpedimentCalculatorByFlag();
    }

    @Test
    void testTimeInImpediment() {
        List<JiraChangelogItem> changelogItems = Arrays.asList(
                new JiraChangelogItem("flagged", null, null, null, null, "impediment", LocalDateTime.of(2019, 1, 1, 12, 0)),
                new JiraChangelogItem("customfield_123", null, null, null, null, null, null),
                new JiraChangelogItem("flagged", null, null, null, null, null, LocalDateTime.of(2019, 1, 10, 12, 0)),
                new JiraChangelogItem("xablau", null, null, null, null, null, null),
                new JiraChangelogItem("flagged", null, null, null, null, "impediment", LocalDateTime.of(2019, 1, 15, 12, 0)),
                new JiraChangelogItem("bla", null, null, null, null, null, null),
                new JiraChangelogItem("other", null, null, null, null, null, null),
                new JiraChangelogItem("flagged", null, null, null, null, null, LocalDateTime.of(2019, 1, 19, 12, 0))
        );

        Assertions.assertThat(calculator.timeInImpediment(changelogItems, LocalDateTime.now(), Collections.emptyList(), Boolean.TRUE))
                .isEqualTo(15);
    }

    @Test
    void testTimeInImpedimentWithoutTerm() {
        List<JiraChangelogItem> changelogItems = Arrays.asList(
                new JiraChangelogItem("flagged", null, null, null, null, "impediment", LocalDateTime.of(2019, 1, 5, 12, 0)),
                new JiraChangelogItem("customfield_123", null, null, null, null, null, null),
                new JiraChangelogItem("flagged", null, null, null, null, null, LocalDateTime.of(2019, 1, 9, 12, 0)),
                new JiraChangelogItem("xablau", null, null, null, null, null, null),
                new JiraChangelogItem("flagged", null, null, null, null, "impediment", LocalDateTime.of(2019, 1, 15, 12, 0)),
                new JiraChangelogItem("bla", null, null, null, null, null, null),
                new JiraChangelogItem("other", null, null, null, null, null, null)
        );

        LocalDateTime endDate = LocalDateTime.of(2019, 1, 19, 12, 0);
        Assertions.assertThat(calculator.timeInImpediment(changelogItems, endDate, Collections.emptyList(), Boolean.TRUE))
                .isEqualTo(10);
    }
}

File: src/test/kotlin/br/com/jiratorio/domain/impediment/calculator/ImpedimentCalculatorByFlagTest.kt
package br.com.jiratorio.domain.impediment.calculator

import br.com.jiratorio.config.junit.testtype.UnitTest
import br.com.jiratorio.domain.FieldChangelog
import br.com.jiratorio.domain.entity.ImpedimentHistory
import br.com.jiratorio.domain.parsed.ParsedChangelog
import br.com.jiratorio.extension.toLocalDateTime
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

@UnitTest
internal class ImpedimentCalculatorByFlagTest {

    @Test
    fun `time in impediment`() {
        val fieldChangelog = setOf(
            FieldChangelog(
                field = "flagged",
                to = "impediment",
                created = "01/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "customfield_123",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "flagged",
                created = "10/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "xablau",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "flagged",
                to = "impediment",
                created = "15/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "other",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "flagged",
                created = "19/01/2019 12:00".toLocalDateTime()
            )
        )

        val impedimentCalculatorResult = ImpedimentCalculatorByFlag.calcImpediment(
            null,
            ParsedChangelog(fieldChangelog.toSet()),
            LocalDateTime.now(),
            emptyList(),
            true
        )

        assertThat(impedimentCalculatorResult.timeInImpediment)
            .isEqualTo(15)
        assertThat(impedimentCalculatorResult.impedimentHistory)
            .hasSize(2)
            .containsExactly(
                ImpedimentHistory(
                    startDate = "01/01/2019 12:00".toLocalDateTime(),
                    endDate = "10/01/2019 12:00".toLocalDateTime(),
                    leadTime = 10
                ),
                ImpedimentHistory(
                    startDate = "15/01/2019 12:00".toLocalDateTime(),
                    endDate = "19/01/2019 12:00".toLocalDateTime(),
                    leadTime = 5
                )
            )
    }

    @Test
    fun `time in impediment without term`() {
        val fieldChangelog = setOf(
            FieldChangelog(
                field = "flagged",
                to = "impediment",
                created = "05/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "customfield_123",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "flagged",
                created = "09/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "xablau",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "flagged",
                to = "impediment",
                created = "15/01/2019 12:00".toLocalDateTime()
            ),
            FieldChangelog(
                field = "bla",
                created = LocalDateTime.now()
            ),
            FieldChangelog(
                field = "other",
                created = LocalDateTime.now()
            )
        )

        val impedimentCalculatorResult = ImpedimentCalculatorByFlag.calcImpediment(
            null,
            ParsedChangelog(fieldChangelog),
            "19/01/2019 12:00".toLocalDateTime(),
            emptyList(),
            true
        )

        assertThat(impedimentCalculatorResult.timeInImpediment)
            .isEqualTo(10)
        assertThat(impedimentCalculatorResult.impedimentHistory)
            .hasSize(2)
            .containsExactly(
                ImpedimentHistory(
                    startDate = "05/01/2019 12:00".toLocalDateTime(),
                    endDate = "09/01/2019 12:00".toLocalDateTime(),
                    leadTime = 5
                ),
                ImpedimentHistory(
                    startDate = "15/01/2019 12:00".toLocalDateTime(),
                    endDate = "19/01/2019 12:00".toLocalDateTime(),
                    leadTime = 5
                )
            )
    }

}
